lock-impl/src/main/java/com/palantir/lock/impl/RateLimitingLockService.java

```diff
@@ -0,0 +1,185 @@
+/**
+ * Copyright 2017 Palantir Technologies
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.lock.impl;
+
+import java.math.BigInteger;
+import java.util.Set;
+
+import javax.annotation.Nullable;
+
+import com.palantir.lock.HeldLocksGrant;
+import com.palantir.lock.HeldLocksToken;
+import com.palantir.lock.LockClient;
+import com.palantir.lock.LockRefreshToken;
+import com.palantir.lock.LockRequest;
+import com.palantir.lock.LockResponse;
+import com.palantir.lock.LockServerOptions;
+import com.palantir.lock.LockService;
+import com.palantir.lock.SimpleHeldLocksToken;
+
+public class RateLimitingLockService implements LockService{
+    private static final ThreadCountLimiter globalLimiter = new ThreadCountLimiter(0);
+
+    private LockService delegate;
+    private ThreadCountLimiter limiter;
+
+    private RateLimitingLockService() {
+        // hidden
+    }
+
+    public static RateLimitingLockService create(LockService delegate, int maxServerThreads, int numClients) {
+        int threadPoolSize = maxServerThreads - 15;
+        int localLimiterSize = threadPoolSize / numClients / 2;
+        int globalLimiterSize = threadPoolSize - localLimiterSize * numClients;
+
+        RateLimitingLockService service = new RateLimitingLockService();
+        service.delegate = delegate;
+        service.limiter = new ThreadCountLimiter(localLimiterSize);
+        globalLimiter.set(globalLimiterSize);
+        return service;
+    }
+
+
+    @Nullable
+    @Override
+    public LockRefreshToken lock(String client, LockRequest request) throws InterruptedException {
+        if (limiter.acquire()) {
+            LockRefreshToken token = delegate.lock(client, request);
+            limiter.release();
+            return token;
+        } else if (globalLimiter.acquire()) {
+            LockRefreshToken token = delegate.lock(client, request);
+            globalLimiter.release();
+            return token;
+        } else {
+            throw new TooManyRequestsException();
+        }
+    }
+
+    @Override
+    public HeldLocksToken lockAndGetHeldLocks(String client, LockRequest request) throws InterruptedException {
+        return null;
+    }
+
+    @Override
+    public boolean unlock(LockRefreshToken token) {
+        return delegate.unlock(token);
+    }
+
+    @Override
+    public Set<LockRefreshToken> refreshLockRefreshTokens(Iterable<LockRefreshToken> tokens) {
```

Comment:
Hmm, does this one need to be rate-limited? I was under the impression that the ones that are uninterruptible should generally be pretty fast (though might be wrong here)